Zsa WIP

PaulLaux opened the pull request:
ZSA wip, changes can be tracked via https://github.com/QED-it/orchard/tree/zsa1

daira commented on src/builder.rs:
```diff
@@ -57,21 +59,23 @@ impl From<value::OverflowError> for Error {
     scope: Scope,
     note: Note,
     merkle_path: MerklePath,
+    // a flag to indicate whether the value of the note will be counted in the value sum of the action.
```
Make sure that the comment reflects the sense of the flag. 
```suggestion
    // If this flag is true, the value of the note will not be counted in the value sum of the action.
```

PaulLaux replied:
done

daira commented on benches/circuit.rs:
```diff
@@ -6,6 +6,7 @@ use criterion::{BenchmarkId, Criterion};
 #[cfg(unix)]
 use pprof::criterion::{Output, PProfProfiler};
 
+use orchard::note::NoteType;
```
Consider renaming to `NoteAsset` ("type" is a very overloaded term).

PaulLaux replied:
We renamed it to be `orchard::note::AssetId` which seems to be more aligned with the spec.

daira commented on src/builder.rs:
```diff
@@ -130,7 +144,17 @@ impl ActionInfo {
     }
 
     /// Returns the value sum for this action.
+    /// Split notes does not contribute to the value sum.
```
```suggestion
    /// Split notes do not contribute to the value sum.
```

daira commented on src/builder.rs:
```diff
@@ -82,16 +86,25 @@ impl SpendInfo {
         }
     }
+
+    /// Duplicates the spend info and set the split flag to `true`.
```
```suggestion
    /// Return a copy of this note with the split flag set to `true`.
```

daira commented on src/builder.rs:
```diff
@@ -130,7 +144,17 @@ impl ActionInfo {
+        //     .spend
+        //     .split_flag
+        //     .then(|| self.spend.note.value())
+        //     .unwrap_or_else(NoteValue::zero);
```
This appears to be the wrong way round: the value should be included iff `split_flag` is false.

PaulLaux replied:
Actually, the code in question does precisely that. I will refactor it as suggested for more clarity.

daira commented on src/builder.rs:
```diff
@@ -140,8 +164,15 @@ impl ActionInfo {
     ///
     /// [orchardsend]: https://zips.z.cash/protocol/nu5.pdf#orchardsend
     fn build(self, mut rng: impl RngCore) -> (Action<SigningMetadata>, Circuit) {
+        assert_eq!(
```
Document the panic condition.

daira commented on src/builder.rs:
```diff
@@ -140,8 +164,15 @@ impl ActionInfo {
     ///
```
```suggestion
    ///
    /// # Panics
    ///
    /// Panics if the asset types of the spent and output notes do not match.
    ///
```

daira commented on src/builder.rs:
```diff
@@ -140,8 +164,15 @@ impl ActionInfo {
     fn build(self, mut rng: impl RngCore) -> (Action<SigningMetadata>, Circuit) {
+        assert_eq!(
+            self.output.note_type,
+            self.spend.note.note_type(),
```
Swap these so that they're in the same order as in the assertion error message.

daira commented on src/builder.rs:
```diff
@@ -151,8 +182,15 @@ impl ActionInfo {
         let ak: SpendValidatingKey = self.spend.fvk.clone().into();
         let alpha = pallas::Scalar::random(&mut rng);
         let rk = ak.randomize(&alpha);
+        let note_type = self.spend.note.note_type();
```
This is unnecessary since we asserted that it's the same as `self.output.note_type` assigned above.
```suggestion
```

daira commented on src/builder.rs:
```diff
@@ -387,6 +442,30 @@ impl Builder {
 }
 
+/// partition a list of spends and recipients by note types.
+fn partition(
```
```suggestion
fn partition_by_asset(
```

PaulLaux replied:
done

daira commented on src/builder.rs:
```diff
@@ -387,6 +442,30 @@ impl Builder {
+) -> HashMap<NoteType, (Vec<SpendInfo>, Vec<RecipientInfo>)> {
+    let mut hm = HashMap::new();
+
+    for s in spends.iter() {
```
```suggestion
    for s in spends {
```

daira commented on src/builder.rs:
```diff
@@ -387,6 +442,30 @@ impl Builder {
+            .push(s.clone());
+    }
+
+    for r in recipients.iter() {
```
```suggestion
    for r in recipients {
```

daira commented on src/constants/fixed_bases.rs:
```diff
@@ -19,8 +19,12 @@ pub mod value_commit_v;
 pub const ORCHARD_PERSONALIZATION: &str = "z.cash:Orchard";
 
 /// SWU hash-to-curve personalization for the value commitment generator
+/// TODO: should we change to "NOTE_TYPE_PERSONALIZATION"?
```
I don't think so; this is still needed with its original value, and the name is still accurate.

PaulLaux replied:
Yeah, make scene, removed it.

daira commented on src/issuance.rs:
```diff
@@ -0,0 +1,731 @@
+//! Structs related to issuance bundles and the associated logic.
```
I have not reviewed this file yet.

daira commented on src/keys.rs:
```diff
@@ -188,21 +188,122 @@ impl SpendValidatingKey {
+/// This type enforces that the corresponding public point (ik^ℙ) has ỹ = 0.
+///
+/// $\mathsf{isk}$ as defined in
+/// [Zcash Protocol Spec § 4.2.3: Orchard Key Components][orchardkeycomponents].
```
Reference the ZSA draft ZIP instead.

PaulLaux replied:
done

daira commented on src/keys.rs:
```diff
@@ -188,21 +188,122 @@ impl SpendValidatingKey {
+///
+/// [orchardkeycomponents]: https://zips.z.cash/protocol/nu5.pdf#orchardkeycomponents
+#[derive(Clone, Debug)]
+pub struct IssuerAuthorizingKey(redpallas::SigningKey<SpendAuth>);
```
This will need to be a separate signature domain.

PaulLaux replied:
I changed `Issuer` to `Issuance` as suggested. Spec changes TBD @daniben31.

`type IssuanceAuth = SpendAuth` still open.

PaulLaux replied:
https://github.com/QED-it/orchard/pull/23/files

PaulLaux replied:
added `type IssuanceAuth = SpendAuth` to be used in all relevant locations.

daira commented on src/keys.rs:
```diff
@@ -188,21 +188,122 @@ impl SpendValidatingKey {
+impl From<&SpendingKey> for IssuerAuthorizingKey {
+    fn from(sk: &SpendingKey) -> Self {
+        let isk = Self::derive_inner(sk);
+        // IssuerSigningKey cannot be constructed such that this assertion would fail.
```
```suggestion
        // IssuerAuthorizingKey cannot be constructed such that this assertion would fail.
```
(or `IssuanceAuthorizingKey`).

PaulLaux replied:
done

daira commented on src/keys.rs:
```diff
@@ -188,21 +188,122 @@ impl SpendValidatingKey {
+
+/// A key used to validate issuer authorization signatures.
+///
+/// Defined in [Zcash Protocol Spec § 4.2.3: Orchard Key Components][orchardkeycomponents].
```
Reference the ZSA draft ZIP.

daira commented on src/keys.rs:
```diff
@@ -188,21 +188,122 @@ impl SpendValidatingKey {
+impl Eq for IssuerValidatingKey {}
+
+impl IssuerValidatingKey {
+    /// Converts this spend validating key to its serialized form,
```
```suggestion
    /// Converts this issuance validating key to its serialized form,
```

PaulLaux replied:
done

daira commented on src/keys.rs:
```diff
@@ -188,21 +188,122 @@ impl SpendValidatingKey {
+
+/// A function to check structural validity of the validating keys for authorizing transfers and
+/// issuing assets
+/// Structural validity checks for ik_P:
```
```suggestion
/// Structural validity checks for ak_P or ik_P:
```

PaulLaux replied:
done

daira commented on src/keys.rs:
```diff
@@ -188,21 +188,122 @@ impl SpendValidatingKey {
+/// issuing assets
+/// Structural validity checks for ik_P:
+///  - The point must not be the identity (which for Pallas is canonically encoded as all-zeroes).
+///  - The sign of the y-coordinate must be positive.
```
```suggestion
///  - The compressed y-coordinate bit must be 0.
```
(This used the wrong terminology in the original code this was moved from.)

daira commented on src/note.rs:
```diff
@@ -86,6 +89,8 @@ pub struct Note {
     /// The value of this note.
     value: NoteValue,
+    /// The type of this note.
+    note_type: NoteType,
```
This could just be `asset: NoteAsset`.

PaulLaux replied:
Thanks for the comments, we are working to make the fixes.

Are you suggesting change every occurrence of `note_type` to `asset` or only this field (which accounts for most of the occurrences) ?

PaulLaux replied:
Ended up with `asset: AssetId` and `note_type` -> `asset` across the entire repo.

daira commented on src/note.rs:
```diff
@@ -151,9 +160,11 @@ impl Note {
     /// Defined in [Zcash Protocol Spec § 4.8.3: Dummy Notes (Orchard)][orcharddummynotes].
     ///
     /// [orcharddummynotes]: https://zips.z.cash/protocol/nu5.pdf#orcharddummynotes
+    /// TODO zsa: remove note_type
```
Why? I think it is necessary to be able to specify the asset when creating a dummy output note, so that it matches the asset of the note being spent in the same action. (There are no non-native dummy spent notes, because a real note with the `split_flag` set is used instead.)

PaulLaux replied:
Right, my initial thinking was that we don't need `dummy_note` for zsa because of split notes. But we do need them for zsa output.

daira commented on src/note.rs:
```diff
@@ -179,6 +191,11 @@ impl Note {
     }
 
+    /// Returns the note type of this note.
+    pub fn note_type(&self) -> NoteType {
```
Again this method can be called `asset()`.

daira commented on src/note/commitment.rs:
```diff
@@ -41,22 +45,46 @@ impl NoteCommitment {
+            // Commit to non-ZEC notes as per the ZSA protocol.
+            // Append the note type to the Orchard note encoding.
+            let type_bits = BitArray::<_, Lsb0>::new(note_type.to_bytes());
+            let zsa_note_bits = zec_note_bits.chain(type_bits.iter().by_vals());
```
If we decide to extend values to 128 bits, this is where the extra high-order bits will be committed.

daira commented on src/note/commitment.rs:
```diff
@@ -41,22 +45,46 @@ impl NoteCommitment {
+        rcm: NoteCommitTrapdoor,
+    ) -> CtOption<Self> {
+        let domain = sinsemilla::CommitDomain::new(personalization);
+        domain.commit(bits, &rcm.0).map(NoteCommitment)
```
If we want this to be constant-time, it will need to also take the maximum number of bits so that it can always do that much computation.

daira commented on src/note/note_type.rs:
```diff
@@ -0,0 +1,113 @@
+        s.extend(ik.to_bytes());
+        s.extend(asset_desc.as_bytes());
+
+        NoteType(assetID_hasher(s))
```
Use a snake-case name, e.g. `asset_id_hasher`, and remove the clippy lint suppression.

daira commented on src/note/note_type.rs:
```diff
@@ -0,0 +1,113 @@
+
+// the hasher used to derive the assetID
+#[allow(non_snake_case)]
+fn assetID_hasher(msg: Vec<u8>) -> pallas::Point {
```
Use a snake-case name, e.g. `asset_id_hasher`, and remove the clippy lint suppression.

PaulLaux replied:
done

daira commented on src/note/note_type.rs:
```diff
@@ -0,0 +1,113 @@
+
+    /// $DeriveNoteType$.
+    ///
+    /// Defined in [Zcash Protocol Spec § TBD: Note Types][notetypes].
```
Reference the ZSA draft ZIP.

PaulLaux replied:
done

daira commented on src/note/note_type.rs:
```diff
@@ -0,0 +1,113 @@
+    /// [notetypes]: https://zips.z.cash/protocol/nu5.pdf#notetypes
+    #[allow(non_snake_case)]
+    pub fn derive(ik: &IssuerValidatingKey, asset_desc: &str) -> Self {
+        assert!(!asset_desc.is_empty() && asset_desc.len() <= MAX_ASSET_DESCRIPTION_SIZE);
```
Document the panic conditions.

PaulLaux replied:
done

daira commented on src/note_encryption.rs:
```diff
@@ -152,12 +177,23 @@ impl Domain for OrchardDomain {
+        np[0] = if is_native { 0x02 } else { 0x03 };
         np[1..12].copy_from_slice(note.recipient().diversifier().as_array());
         np[12..20].copy_from_slice(&note.value().to_bytes());
+        // todo: add note_type
```
Isn't that done just below?
```suggestion
```

daira commented on src/note_encryption.rs:
```diff
@@ -224,9 +260,20 @@ impl Domain for OrchardDomain {
+                memo.copy_from_slice(full_memo);
+            }
+            0x03 => {
+                // ZSA note plaintext have a shorter memo.
```
```suggestion
                // ZSA note plaintexts have a shorter memo.
```

daira commented on src/note_encryption.rs:
```diff
@@ -254,6 +301,10 @@ impl BatchDomain for OrchardDomain {
     }
 }
 
+fn get_note_version(plaintext: &NotePlaintextBytes) -> u8 {
```
```suggestion
fn get_note_plaintext_version(plaintext: &NotePlaintextBytes) -> u8 {
```

PaulLaux commented:
We addressed most issues in https://github.com/QED-it/orchard/pull/23 and https://github.com/QED-it/orchard/pull/24 the rest are tagged as in-progress.

Closing in favor of the updated https://github.com/zcash/orchard/pull/364.
